engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/state/RunStateNode.java

```diff
@@ -1,18 +1,163 @@
+  abstract Stateful execute(
+      @MonadicState Object state, Object _this, Object key, Object local_state, Thunk computation);
+
+  @Specialization
+  Stateful doEmpty(
+      EmptyMap state, Object _this, Object key, Object local_state, Thunk computation) {
+    SingletonMap localStateMap = new SingletonMap(key, local_state);
+    Object result = thunkExecutorNode.executeThunk(computation, localStateMap, false).getValue();
+    return new Stateful(state, result);
+  }
+
+  @Specialization(guards = {"state.getKey() == key"})
+  Stateful doSingletonSameKey(
+      SingletonMap state, Object _this, Object key, Object local_state, Thunk computation) {
+    SingletonMap localStateContainer = new SingletonMap(state.getKey(), local_state);
+    Stateful res = thunkExecutorNode.executeThunk(computation, localStateContainer, false);
+    return new Stateful(state, res.getValue());
+  }
+
+  @Specialization(
+      guards = {
+        "key == cachedNewKey",
+        "state.getKey() == cachedOldKey",
+        "cachedOldKey != cachedNewKey"
+      })
+  Stateful doSingletonNewKeyCached(
+      SingletonMap state,
+      Object _this,
+      Object key,
+      Object local_state,
+      Thunk computation,
+      @Cached("key") Object cachedNewKey,
+      @Cached("state.getKey()") Object cachedOldKey,
+      @Cached(value = "buildSmallKeys(cachedNewKey, cachedOldKey)", dimensions = 1)
+          Object[] newKeys) {
+    SmallMap localStateMap = new SmallMap(newKeys, new Object[] {local_state, state.getValue()});
+    Stateful res = thunkExecutorNode.executeThunk(computation, localStateMap, false);
+    Object newStateVal = ((SmallMap) res.getState()).getValues()[1];
+    return new Stateful(new SingletonMap(cachedOldKey, newStateVal), res.getValue());
+  }
+
+  @Specialization
+  Stateful doSingletonNewKeyUncached(
+      SingletonMap state, Object _this, Object key, Object local_state, Thunk computation) {
+    return doSingletonNewKeyCached(
+        state,
+        _this,
+        key,
+        local_state,
+        computation,
+        key,
+        state.getKey(),
+        buildSmallKeys(key, state.getKey()));
+  }
+
+  Object[] buildSmallKeys(Object k1, Object k2) {
+    return new Object[] {k1, k2};
+  }
+
+  @Specialization(
+      guards = {"key == cachedNewKey", "state.getKeys() == cachedOldKeys", "index == NOT_FOUND"})
+  Stateful doMultiNewKeyCached(
+      SmallMap state,
+      Object _this,
+      Object key,
+      Object local_state,
+      Thunk computation,
+      @Cached("key") Object cachedNewKey,
+      @Cached(value = "state.getKeys()", dimensions = 1) Object[] cachedOldKeys,
+      @Cached("state.indexOf(key)") int index,
+      @Cached(value = "buildNewKeys(cachedNewKey, cachedOldKeys)", dimensions = 1)
+          Object[] newKeys) {
+    Object[] newValues = new Object[newKeys.length];
+    System.arraycopy(state.getValues(), 0, newValues, 1, cachedOldKeys.length);
+    newValues[0] = local_state;
```

Comment:
Maybe this could be extracted as an `Object[] prependArray(Object[] array, Object obj)`?

As long as using a helper function does not hurt performance, I think it would be much easier to understand this code if the `System.arraycopy` logic was done in low-level helper functions and the overall state manipulation was one level of abstraction higher.